### PR TITLE
feat(cats/sql): account name based sharding logic

### PR DESCRIPTION
**Pull request by kirangodishala**
Addressing this issue - #[6315](https://github.com/spinnaker/spinnaker/issues/6315)

This feature can be enabled by setting `caching.sharding-enabled `property to `true`. No other configuration is needed.

When this feature is enabled, caching of accounts will be shared across the available pods in such a way that agents of an account are always run by one particular pod determined by a hashing logic. When the pods are scaled up or down, the agents will resettle to the same or a different pod based on the hashing logic. 

This implementation is for the SQL backed clouddriver(planning to extend the same to redis module as well).

### CODE REVIEW

**Review by german-muzquiz — COMMENTED** [review]
I only have very minor comments, the implementation looks solid!

File: cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cluster/SqlCachingPodsObserver.kt [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        val existingReplicas = jooq.select()
+          .from(table(replicasTable))
+          .fetch()
+          .intoResultSet()
+        val now = System.currentTimeMillis()
+        while (existingReplicas.next()) {
+          val expiry = existingReplicas.getLong(LAST_HEARTBEAT_TIME)
+          if (now > expiry) {
+            try {
+              jooq.deleteFrom(table(replicasTable))
```
[the remark this answers is not in the document]
Comment (kirangodishala) [reply cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cluster/SqlCachingPodsObserver.kt:140]: log entry added.

File: cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cluster/SqlCachingPodsObserver.kt [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        }
+      }
+    }catch (e: SQLException){
+      log.error( "Failed to fetch live pods count ${e.message}")
+    }
+    if(counter == 0 || index == -1){
+      throw RuntimeException("No caching pod heartbeat records detected. Sharding logic can't be applied!!!!")
+    }
+    podCount = counter
+    podIndex = index
```
[the remark this answers is not in the document]
Comment (kirangodishala) [reply cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cluster/SqlCachingPodsObserver.kt:190]: debug level log added.

File: cats/cats-core/src/main/java/com/netflix/spinnaker/cats/cluster/ShardingFilter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.cats.cluster;
+
+import com.netflix.spinnaker.cats.agent.Agent;
+
+public interface ShardingFilter {
+  Boolean filter(Agent agent);
```
[the remark this answers is not in the document]
Comment (kirangodishala) [reply cats/cats-core/src/main/java/com/netflix/spinnaker/cats/cluster/ShardingFilter.java]: done

**Review by german-muzquiz — APPROVED** [review]

**Review by clanesf — APPROVED** [review]
Looks great. Excited to use it!

**Pull request merged**